File: pulse-kotlin/src/commonMain/kotlin/com/fluxchess/pulse/kotlin/uci/Receiver.kt
/*
 * Copyright 2013-2024 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package com.fluxchess.pulse.kotlin.uci

import com.fluxchess.pulse.kotlin.engine.MoveList
import com.fluxchess.pulse.kotlin.engine.generateLegalMoves

const val STARTING_POSITION_FEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

private val uciTokenRegex = Regex("\\s+")

private val nameValueOptionRegex = Regex("^name\\s+(?<name>.+?)\\s+value\\s+(?<value>.+)")
private val nameOnlyOptionRegex = Regex("^name\\s+(?<name>.+)")

private val startposMovesRegex = Regex("^startpos\\s+moves\\s+(?<moves>.+)")
private val fenMovesRegex = Regex("^fen\\s+(?<fen>.+?)\\s+moves\\s+(?<moves>.+)")
private val startposOnlyRegex = Regex("^startpos")
private val fenOnlyRegex = Regex("^fen\\s+(?<fen>.+)")

private val movesTokenRegex = Regex("\\s+")

interface Receiver {
	fun run()
}

class DefaultReceiver(
	private val reader: Reader,
	private val sender: DefaultSender,
	private val engine: Engine,
) : Receiver {
	override fun run() {
		while (true) {
			val line = reader.readln()
			if (line == null) {
				engine.quit()
				return
			}
			val tokens = line.trim().split(uciTokenRegex, 2)
			if (tokens.isEmpty()) {
				continue
			}
			when (tokens[0]) {
				"uci" -> engine.initialize()
				"debug" -> parseDebug(tokens)
				"isready" -> engine.ready()
				"setoption" -> parseSetOption(tokens)
				"register" -> sender.debug("Unsupported command: register")
				"ucinewgame" -> engine.newGame()
				"position" -> parsePosition(tokens)
				"go" -> parseGo()
				"stop" -> engine.stop()
				"ponderhit" -> engine.ponderHit()
				"quit" -> {
					engine.quit()
					return
				}

				else -> sender.debug("Unknown command: ${tokens[0]}")
			}
		}
	}

	private fun parseDebug(tokens: List<String>) {
		when (tokens.size) {
			1 -> sender.debugMode = !sender.debugMode // Toggle debug
			2 -> when {
				tokens[1] == "on" -> sender.debugMode = true
				tokens[1] == "off" -> sender.debugMode = false
				else -> sender.debug("Unknown argument: ${tokens[1]}")
			}
		}
	}

	private fun parseSetOption(tokens: List<String>) {
		if (tokens.size != 2) {
			sender.debug("Argument required")
			return
		}
		val nameValueMatch = nameValueOptionRegex.find(tokens[1])
		if (nameValueMatch != null) {
			val (name, value) = nameValueMatch.destructured
			engine.setOption(name, value)
			return
		}
		val nameOnlyMatch = nameOnlyOptionRegex.find(tokens[1])
		if (nameOnlyMatch != null) {
			val (name) = nameOnlyMatch.destructured
			engine.setOption(name)
			return
		}
		sender.debug("Error parsing argument: ${tokens[1]}")
	}

	private fun parsePosition(tokens: List<String>) {
		if (tokens.size != 2) {
			sender.debug("Argument required")
			return
		}
		val startposMovesMatch = startposMovesRegex.find(tokens[1])
		if (startposMovesMatch != null) {
			val (moves) = startposMovesMatch.destructured
			playMoves(STARTING_POSITION_FEN, moves)
			return
		}
		val fenMovesMatch = fenMovesRegex.find(tokens[1])
		if (fenMovesMatch != null) {
			val (fen, moves) = fenMovesMatch.destructured
			playMoves(fen, moves)
			return
		}
		val startposOnlyMatch = startposOnlyRegex.find(tokens[1])
		if (startposOnlyMatch != null) {
			playMoves(STARTING_POSITION_FEN)
			return
		}
		val fenOnlyMatch = fenOnlyRegex.find(tokens[1])
		if (fenOnlyMatch != null) {
			val (fen) = fenOnlyMatch.destructured
			playMoves(fen)
			return
		}
		sender.debug("Error parsing argument: ${tokens[1]}")
	}

	private fun playMoves(fen: String, moves: String = "") {
		val legalMoves = MoveList()
		val position = fen.toPositionOrNull()
		if (position == null) {
			sender.debug("Invalid position: $fen")
			return
		}
		if (moves.isNotBlank()) {
			val moveList = moves.split(movesTokenRegex)
			moveList.forEach { move ->
				generateLegalMoves(legalMoves, position)
				for (i in 0 until legalMoves.size) {
					val legalMove = legalMoves.entries[i].move
					if (legalMove.toNotation() == move) {
						position.makeMove(legalMove)
						return@forEach
					}
				}
				sender.debug("Invalid move: $move, position: $position")
				return
			}
		}
		engine.position(position)
		return
	}

	private fun parseGo() {
	}
}

File: pulse-go/internal/pulse/uci/receiver.go
/*
 * Copyright 2013-2023 Phokham Nonava
 *
 * Use of this source code is governed by the MIT license that can be
 * found in the LICENSE file.
 */

package uci

import (
	"fmt"
	"io"
	"regexp"
	"strings"

	"github.com/fluxroot/pulse/internal/pulse/engine"
)

const startingPositionFEN = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

var uciTokenRegex = regexp.MustCompile(`\s+`)

var nameValueOptionRegex = regexp.MustCompile(`^name\s+(?P<name>.+?)\s+value\s+(?P<value>.+)`)
var nameOnlyOptionRegex = regexp.MustCompile(`^name\s+(?P<name>.+)`)

var startposMovesRegex = regexp.MustCompile(`^startpos\s+moves\s+(?P<moves>.+)`)
var fenMovesRegex = regexp.MustCompile(`^fen\s+(?P<fen>.+?)\s+moves\s+(?P<moves>.+)`)
var startposOnlyRegex = regexp.MustCompile(`^startpos`)
var fenOnlyRegex = regexp.MustCompile(`^fen\s+(?P<fen>.+)`)

var movesTokenRegex = regexp.MustCompile(`\s+`)

type Receiver interface {
	Run() error
}

func NewDefaultReceiver(reader Reader, sender *DefaultSender, engine Engine) *DefaultReceiver {
	return &DefaultReceiver{
		reader: reader,
		sender: sender,
		engine: engine,
	}
}

type DefaultReceiver struct {
	reader Reader
	sender *DefaultSender
	engine Engine
}

func (r *DefaultReceiver) Run() error {
	for {
		line, err := r.reader.Readln()
		if err == io.EOF {
			_ = r.engine.Quit()
			return nil
		}
		if err != nil {
			_ = r.engine.Quit()
			return err
		}
		tokens := uciTokenRegex.Split(strings.TrimSpace(line), 2)
		if len(tokens) == 0 {
			continue
		}
		switch tokens[0] {
		case "uci":
			if err := r.engine.Initialize(); err != nil {
				return fmt.Errorf("initialize: %w", err)
			}
		case "debug":
			r.parseDebug(tokens)
		case "isready":
			if err := r.engine.Ready(); err != nil {
				return fmt.Errorf("ready: %w", err)
			}
		case "setoption":
			r.parseSetOption(tokens)
		case "register":
			if err := r.sender.Debug("Unsupported command: register"); err != nil {
				return fmt.Errorf("debug: %w", err)
			}
		case "ucinewgame":
			if err := r.engine.NewGame(); err != nil {
				return fmt.Errorf("new game: %w", err)
			}
		case "position":
			r.parsePosition(tokens)
		case "go":
			parseGo()
		case "stop":
			if err := r.engine.Stop(); err != nil {
				return fmt.Errorf("stop: %w", err)
			}
		case "ponderhit":
			if err := r.engine.PonderHit(); err != nil {
				return fmt.Errorf("ponder hit: %w", err)
			}
		case "quit":
			if err := r.engine.Quit(); err != nil {
				return fmt.Errorf("quit: %w", err)
			}
			return nil
		default:
			if err := r.sender.Debug(fmt.Sprintf("Unknown command: %s", tokens[0])); err != nil {
				return fmt.Errorf("debug: %w", err)
			}
		}
	}
}

func (r *DefaultReceiver) parseDebug(tokens []string) {
	switch len(tokens) {
	case 1:
		r.sender.debugMode = !r.sender.debugMode // Toggle debug
	case 2:
		switch tokens[1] {
		case "on":
			r.sender.debugMode = true
		case "off":
			r.sender.debugMode = false
		default:
			_ = r.sender.Debug(fmt.Sprintf("Unknown argument: %s", tokens[1]))
		}
	}
}

func (r *DefaultReceiver) parseSetOption(tokens []string) {
	if len(tokens) != 2 {
		_ = r.sender.Debug("Argument required")
		return
	}
	nameValueMatch := nameValueOptionRegex.FindStringSubmatch(tokens[1])
	if nameValueMatch != nil {
		result := map[string]string{}
		for i, name := range nameValueOptionRegex.SubexpNames() {
			if i != 0 && name != "" {
				result[name] = nameValueMatch[i]
			}
		}
		r.engine.SetNameValueOption(result["name"], result["value"])
		return
	}
	nameOnlyMatch := nameOnlyOptionRegex.FindStringSubmatch(tokens[1])
	if nameOnlyMatch != nil {
		result := map[string]string{}
		for i, name := range nameOnlyOptionRegex.SubexpNames() {
			if i != 0 && name != "" {
				result[name] = nameOnlyMatch[i]
			}
		}
		r.engine.SetNameOnlyOption(result["name"])
		return
	}
	_ = r.sender.Debug(fmt.Sprintf("Error parsing argument: %s", tokens[1]))
}

func (r *DefaultReceiver) parsePosition(tokens []string) {
	if len(tokens) != 2 {
		_ = r.sender.Debug("Argument required")
		return
	}
	startposMovesMatch := startposMovesRegex.FindStringSubmatch(tokens[1])
	if startposMovesMatch != nil {
		result := map[string]string{}
		for i, name := range startposMovesRegex.SubexpNames() {
			if i != 0 && name != "" {
				result[name] = startposMovesMatch[i]
			}
		}
		r.playMoves(startingPositionFEN, result["moves"])
		return
	}
	fenMovesMatch := fenMovesRegex.FindStringSubmatch(tokens[1])
	if fenMovesMatch != nil {
		result := map[string]string{}
		for i, name := range fenMovesRegex.SubexpNames() {
			if i != 0 && name != "" {
				result[name] = fenMovesMatch[i]
			}
		}
		r.playMoves(result["fen"], result["moves"])
		return
	}
	startposOnlyMatch := startposOnlyRegex.FindStringSubmatch(tokens[1])
	if startposOnlyMatch != nil {
		r.playMoves(startingPositionFEN, "")
		return
	}
	fenOnlyMatch := fenOnlyRegex.FindStringSubmatch(tokens[1])
	if fenOnlyMatch != nil {
		result := map[string]string{}
		for i, name := range fenOnlyRegex.SubexpNames() {
			if i != 0 && name != "" {
				result[name] = fenOnlyMatch[i]
			}
		}
		r.playMoves(result["fen"], "")
		return
	}
	_ = r.sender.Debug(fmt.Sprintf("Error parsing argument: %s", tokens[1]))
}

func (r *DefaultReceiver) playMoves(fen string, moves string) {
	legalMoves := &engine.MoveList{}
	position, err := FENToPosition(fen)
	if err != nil {
		_ = r.sender.Debug(fmt.Sprintf("Invalid position: %s", fen))
		return
	}
	if moves != "" {
		moveList := movesTokenRegex.Split(moves, -1)
	MoveList:
		for _, m := range moveList {
			engine.GenerateLegalMoves(legalMoves, position)
			for i := 0; i < legalMoves.Size; i++ {
				legalMove := legalMoves.Entries[i].Move
				if MoveToNotation(legalMove) == m {
					position.MakeMove(legalMove)
					continue MoveList
				}
			}
			_ = r.sender.Debug(fmt.Sprintf("Invalid move: %s, position: %s", m, PositionToFEN(position)))
			return
		}
	}
	r.engine.Position(position)
	return
}

func parseGo() {
}
